### PR TITLE
Implement duplicating existing entry

### PR DESCRIPTION
**Pull request by mrzhdev**
## Description
Feature for creating new bookings from existing ones.
Added  a context menu for booking entries.

## Related issue
#187

### CODE REVIEW

File: src/SimpleAccounting/Model/ProjectData.cs [hunk 1]
```diff
@@ -227,37 +227,13 @@ public void SelectYear(string yearName)
             this.YearChanged(this, EventArgs.Empty);
         }
 
-        public void ShowAddBookingDialog(bool showInactiveAccounts)
+        private EditBookingViewModel BookingFromJournal(AccountingDataJournalBooking journalEntry, bool isEdit, bool showInactiveAccounts)
```
Comment (lg2de) [inline src/SimpleAccounting/Model/ProjectData.cs:230]: `private` methods should be placed after `public` (and `protected`) methods.
If you have "ReSharper" or "Rider" please try the cleanup profile "lg2".

File: src/SimpleAccounting/Model/ProjectData.cs [hunk 2]
```diff
@@ -290,6 +266,52 @@ public void ShowEditBookingDialog(IJournalItem item, bool showInactiveAccounts)
             var allAccounts = this.Storage.AllAccounts;
             bookingModel.Accounts.AddRange(showInactiveAccounts ? allAccounts : allAccounts.Where(x => x.Active));
 
+            return bookingModel;
+
```
Comment (lg2de) [inline src/SimpleAccounting/Model/ProjectData.cs:270]:
```suggestion
```

File: src/SimpleAccounting/Model/ProjectData.cs (hunk 1 above)
Comment (lg2de) [inline src/SimpleAccounting/Model/ProjectData.cs:230]:
```suggestion
        private EditBookingViewModel CreateEditViewModel(AccountingDataJournalBooking journalEntry, bool editMode, bool showInactiveAccounts)
```

File: src/SimpleAccounting/Model/ProjectData.cs [hunk 3]
```diff
@@ -227,37 +227,13 @@ public void SelectYear(string yearName)
...
-                return;
-            }
-
-            var journalEntry = this.CurrentYear.Booking[item.StorageIndex];
-
             var bookingModel = new EditBookingViewModel(
                 this,
                 journalEntry.Date.ToDateTime(),
-                editMode: true)
+                editMode: isEdit)
```
Comment (lg2de) [inline src/SimpleAccounting/Model/ProjectData.cs:235]:
```suggestion
                editMode)
```

File: src/SimpleAccounting/Model/ProjectData.cs [hunk 4]
```diff
@@ -290,6 +266,52 @@ public void ShowEditBookingDialog(IJournalItem item, bool showInactiveAccounts)
...
+        public void ShowDuplicateBookingDialog(IJournalItem item, bool showInactiveAccounts)
+        {
+            if (item.StorageIndex < 0)
+            {
+                // summary item selected => ignore
+                return;
+            }
+            var journalEntry = this.CurrentYear.Booking[item.StorageIndex];
+
+            var bookingModel = this.BookingFromJournal(journalEntry, false, showInactiveAccounts);
```
Comment (lg2de) [inline src/SimpleAccounting/Model/ProjectData.cs:297]:
```suggestion
            var bookingModel = this.BookingFromJournal(journalEntry, editMode: false, showInactiveAccounts); 
```

File: src/SimpleAccounting/Model/ProjectData.cs [hunk 5]
```diff
@@ -290,6 +266,52 @@ public void ShowEditBookingDialog(IJournalItem item, bool showInactiveAccounts)
...
+        {
+            if (item.StorageIndex < 0)
+            {
+                // summary item selected => ignore
+                return;
+            }
+
+            var journalEntry = this.CurrentYear.Booking[item.StorageIndex];
+
+            var bookingModel = this.BookingFromJournal(journalEntry, true, showInactiveAccounts);
```
Comment (lg2de) [inline src/SimpleAccounting/Model/ProjectData.cs:312]:
```suggestion
            var bookingModel = this.BookingFromJournal(journalEntry, editMode: true, showInactiveAccounts);
```

File: src/SimpleAccounting/Properties/Resources.de.resx [hunk 6]
```diff
@@ -579,4 +579,7 @@ Ohne eine solche Konfiguration für mindestens ein Konto ist diese Funktion deak
   <data name="AccountType_Carryforward" xml:space="preserve">
     <value>Saldovortrag</value>
   </data>
+  <data name="JournalContextMenu_Duplicate" xml:space="preserve">
+    <value>Duplikat</value>
```
Comment (lg2de) [inline src/SimpleAccounting/Properties/Resources.de.resx:583]:
```suggestion
    <value>Kopie erstellen</value>
```

File: src/SimpleAccounting/Properties/Resources.resx [hunk 7]
```diff
@@ -594,4 +594,7 @@ Without such configuration for at least one account, this function is disabled.<
   <data name="AccountType_Carryforward" xml:space="preserve">
     <value>Carry forward</value>
   </data>
+  <data name="JournalContextMenu_Duplicate" xml:space="preserve">
+    <value>Duplicate</value>
```
Comment (lg2de) [inline src/SimpleAccounting/Properties/Resources.resx:598]:
```suggestion
    <value>Create copy</value>
```

Comment (lg2de) [reply src/SimpleAccounting/Properties/Resources.resx:598]: Please change french version accordingly.

File: src/SimpleAccounting/Presentation/ShellView.xaml [hunk 8]
```diff
@@ -174,6 +174,13 @@
                 Command="{Binding Menu.EditBookingCommand}"
                 CommandParameter="{Binding FullJournal.Items/}" />
             </DataGrid.InputBindings>
+            <DataGrid.ContextMenu>
```
Comment (lg2de) [inline src/SimpleAccounting/Presentation/ShellView.xaml:177]: When starting context menu, could you please also add the "edit" option?

File: src/SimpleAccounting/Model/ProjectData.cs [hunk 9]
```diff
@@ -290,6 +266,52 @@ public void ShowEditBookingDialog(IJournalItem item, bool showInactiveAccounts)
...
+        }
+
+        public void ShowDuplicateBookingDialog(IJournalItem item, bool showInactiveAccounts)
+        {
+            if (item.StorageIndex < 0)
+            {
+                // summary item selected => ignore
+                return;
+            }
+            var journalEntry = this.CurrentYear.Booking[item.StorageIndex];
```
Comment (lg2de) [inline src/SimpleAccounting/Model/ProjectData.cs:295]:
```suggestion

            var journalEntry = this.CurrentYear.Booking[item.StorageIndex];
```

**Review by lg2de — CHANGES_REQUESTED** [review]
Thanks for you contribution and sorry for the delay!
Please also add a unit test that verifies that duplicating an entry will not change the old one.

**Comment by lg2de** [comment]
@mrzhdev Are you still willing to implement my suggestions?

File: src/SimpleAccounting/Properties/Resources.resx (hunk 7 above)
Comment (mrzhdev) [reply src/SimpleAccounting/Properties/Resources.resx:598]: Créer une copie

**Comment by mrzhdev** [comment]
I'm quite busy these days and I do not have the efficiency to implement
your suggestions in reasonable delay.

**Pull request merged**